### PR TITLE
Admin API endpoint to delete a reported event

### LINKED ISSUE #15011: Admin API endpoint to delete a reported event
**Description:**

Currently the [event reports API](https://matrix-org.github.io/synapse/latest/admin_api/event_reports.html) can be used to list all reported events stored in a synapse instance.

While this is enough to browse events, AFAIK there is no way to delete elements from this list of reported events, eg. to delete  a report considered invalid by an admin or by the report author.

I would expect being able to do something like:

```
DELETE /_synapse/admin/v1/event_reports/<event_report_id>
```

### PR DESCRIPTION
**Pull request by dklimpel**
Closes #15011

Add admin API endpoint to delete a reported event.

~~Move `get_event_report` and `get_event_reports_paginate` from `RoomStore` to `RoomWorkerStore`.
Both are read only queries. IMO there is no reason why this is should not possible for workers.~~

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel [5740567+dklimpel@users.noreply.github.com](mailto:5740567+dklimpel@users.noreply.github.com)

### CODE REVIEW

**Comment by H-Shay** [comment]
Looks like there is a merge conflict, would you mind resolving that before review?

File: tests/rest/admin/test_event_reports.py [hunk 1]
```diff
@@ -599,3 +599,133 @@ def _check_fields(self, content: JsonDict) -> None:
...
+                "event_id",
+                self.other_user,
+                "this makes me sad",
+                {},
+                self.clock.time_msec(),
+            )
+        )
+
+        # first created event report gets `id`=2
+        self.url = "/_synapse/admin/v1/event_reports/2"
```
Comment (DMRobertson) [inline tests/rest/admin/test_event_reports.py:632]: Can we avoid hard coding this by making `add_event_report` return `next_id` and reading that value here?

File: tests/rest/admin/test_event_reports.py [hunk 2]
```diff
@@ -599,3 +599,133 @@ def _check_fields(self, content: JsonDict) -> None:
...
+        """
+
+        channel = self.make_request(
+            "DELETE",
+            self.url,
+            access_token=self.admin_user_tok,
+        )
+
+        self.assertEqual(200, channel.code, msg=channel.json_body)
+        self.assertEqual({}, channel.json_body)
```
Comment (DMRobertson) [inline tests/rest/admin/test_event_reports.py:669]: This test should check that the delete actually took place in the database by trying to fetch it again, e.g. by calling `RoomStore.get_event_report`. Alternatively you could make an admin API request to fetch the event and check that it fails in a sensible way.

**Review by DMRobertson — CHANGES_REQUESTED** [review]
> Move get_event_report and get_event_reports_paginate from RoomStore to RoomWorkerStore.
Broadly LGTM, but I've left a few notes.

> Both are read only queries. IMO there is no reason why this is should not possible for workers.

Agreed on this, but if possible could you try to keep this to a separate commit (or even PR) in future? It's helpful for the reviewer (if independent changes are separate.)

**Comment by dklimpel** [comment]
> Agreed on this, but if possible could you try to keep this to a separate commit (or even PR) in future? It's helpful for the reviewer (if independent changes are separate.)

I have moved it to #15165

File: synapse/storage/databases/main/room.py [hunk 3]
```diff
@@ -2160,7 +2160,19 @@ async def add_event_report(
...
+    ) -> int:
+        """Add an event report
+
+        Args:
+            room_id: Room that contains the reported event.
+            event_id: The reported event.
+            user_id: User who reports the event.
+            reason: Description that the user specifies.
+            content: Report request body (score and reason).
+            received_ts: Time when the user submitted the report.
```
Comment (DMRobertson) [inline synapse/storage/databases/main/room.py:2172]:
```suggestion
            received_ts: Time when the user submitted the report (milliseconds).
```

**Review by DMRobertson — APPROVED** [review]
Thanks Dirk, LGTM!

**Pull request merged**